### PR TITLE
[Feature, Example] A3C Atari Implementation for TorchRL

### PR DESCRIPTION
**Pull request by simeetnayan81**
## Description

Describe your changes in detail.
This PR adds an implementation of the Asynchronous Advantage Actor-Critic (A3C) algorithm for Atari environments in the [torchrl/sota-implementations](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) directory. The main files added are:

a3c_atari.py: Contains the A3C worker class, shared optimizer, and main training loop using multiprocessing.
utils_atari.py: Provides utility functions for environment creation, model construction, and evaluation, adapted for Atari tasks.
config_atari.yaml: Configuration file for hyperparameters, environment settings, and logging.

The implementation leverages TorchRL's collectors, objectives, and logging utilities, and is designed to be modular and extensible for research and benchmarking. Some of the utils functions are also borrowed from a2c_atari.

## Motivation and Context

This change is required to provide a strong, reproducible baseline for A3C on Atari environments using TorchRL. It enables researchers and practitioners to benchmark and compare reinforcement learning algorithms within the TorchRL ecosystem. The implementation follows best practices for distributed RL and is compatible with TorchRL's API.

This PR solves the issue: #1755

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vmoens — COMMENTED** [review]
This all looks pretty good!
Could you share a (couple of) learning curve?
Another thing to do before landing is to add it to the sota-implementations CI run:
https://github.com/pytorch/rl/blob/main/.github/unittest/linux_sota/scripts/test_sota.py
Make sure the config passed there is as much barebone as we can - we just want to run the script for a couple of collection / optim iters and make sure it runs without error (not that it properly trains).
We also need to add it to the [sota-check](https://github.com/pytorch/rl/tree/main/sota-check) runs

**Comment by simeetnayan81** [comment]
Thanks @vmoens . I'll add the required changes as well as some training curves.

**Comment by simeetnayan81** [comment]
@vmoens, I have added the required scripts as well.

Not getting enough resources and time for hyperparam tuning to generate a proper training curve.

File: sota-implementations/a3c/a3c_atari.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        logger = get_logger(
+            cfg.logger.backend,
+            logger_name="a3c",
+            experiment_name=exp_name,
+            wandb_kwargs={
+                "config": dict(cfg),
+                "project": cfg.logger.project_name,
+                "group": cfg.logger.group_name,
+            },
+        )
```
Comment (vmoens) [inline sota-implementations/a3c/a3c_atari.py]: What I usually see is that the logger is only passed to the first worker. 
Another thing is that you may want to assume that the logger isn't serializable and should be instantiated locally within the worker.

**Review by vmoens — COMMENTED** [review]
LGTM, just a minor comment on the logger!

File: sota-implementations/a3c/a3c_atari.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        cfg.env.env_name, device=cfg.loss.device, gym_backend=cfg.env.backend
+    )
+    global_model = nn.ModuleList([global_actor, global_critic_head])
+    global_model.share_memory()
+    optimizer = SharedAdam(global_model.parameters(), lr=cfg.optim.lr)
+
+    num_workers = cfg.multiprocessing.num_workers
+
+    if num_workers is None:
+        num_workers = mp.cpu_count()
```
Comment (vmoens) [inline sota-implementations/a3c/a3c_atari.py]: we should have way fewer workers - I think we need users to tell us how many.

File: sota-implementations/a3c/a3c_atari.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                metrics_to_log[
+                    "train/episode_length"
+                ] = episode_length.sum().item() / len(episode_length)
+
+            with torch.no_grad():
+                data = self.adv_module(data)
+            data_reshape = data.reshape(-1)
+            losses = []
+
+            mini_batches = data_reshape.split(self.mini_batch_size)
```
Comment (vmoens) [inline sota-implementations/a3c/a3c_atari.py:198]: To shuffle things a bit I usually rely on a replay buffer instance rather than just splitting the data

File: sota-implementations/a3c/a3c_atari.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            global_param._grad = local_param.grad
+
+        for local_param, global_param in zip(
+            self.local_critic.parameters(), self.global_critic.parameters()
+        ):
+            global_param._grad = local_param.grad
+
+        gn = torch.nn.utils.clip_grad_norm_(
+            self.loss_module.parameters(), max_norm=max_grad_norm
+        )
```
Comment (vmoens) [inline sota-implementations/a3c/a3c_atari.py:134]: can you explain what we do here? What do we use the _grad for?

File: sota-implementations/a3c/a3c_atari.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+from torchrl.objectives.value.advantages import GAE
+
+from torchrl.record.loggers import generate_exp_name, get_logger
+from utils_atari import make_parallel_env, make_ppo_models
+
+
+torch.set_float32_matmul_precision("high")
+
+
+class SharedAdam(torch.optim.Adam):
```
Comment (vmoens) [inline sota-implementations/a3c/a3c_atari.py]: shouldn't we move this to the utils file?

**Review by vmoens — COMMENTED** [review]
I made a few edits.
Can you explain the way the params are shared and updated? I'm not sure I see the logic

File: sota-implementations/a3c/a3c_atari.py (hunk 4 above)
Comment (simeetnayan81) [reply sota-implementations/a3c/a3c_atari.py:134]: _grad is used to store the gradients for each parameter.
We copy local gradients to the global model so the global model can be updated with the optimizer.
This is a key step in A3C, where multiple workers asynchronously update a shared global model.

File: sota-implementations/a3c/a3c_atari.py (hunk 5 above)
Comment (simeetnayan81) [reply sota-implementations/a3c/a3c_atari.py]: Sure, will do it

File: sota-implementations/a3c/a3c_atari.py (hunk 2 above)
Comment (simeetnayan81) [reply sota-implementations/a3c/a3c_atari.py]: That can be configured in the config_atari. You want me to explicitly set it to some constant here?

File: sota-implementations/a3c/a3c_atari.py (hunk 1 above)
Comment (simeetnayan81) [reply sota-implementations/a3c/a3c_atari.py]: Oh yea, I did that because I thought logging any single worker should be a good representative of the global model since anyway the weights are being copied. Logging all the worker might not be really useful but that can be done as well.

**Comment by simeetnayan81** [comment]
> I made a few edits. Can you explain the way the params are shared and updated? I'm not sure I see the logic

There is a global model (shared across all workers) and a local model (each worker has its own copy).
The global model’s parameters are placed in shared memory so all workers can access and update them.
Each worker interacts with its own environment using its local model.
After collecting experience, the worker computes gradients (via backpropagation) on its local model.
The gradients from the local model are copied to the corresponding parameters of the global model (global_param._grad = local_param.grad).
The optimizer (e.g., Adam) is called on the global model, using the gradients just copied from the local model.
This updates the global model’s parameters.
The local model can be periodically updated (synced) from the global model to keep it up-to-date.

File: sota-implementations/a3c/a3c_atari.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        for local_param, global_param in zip(
+            self.local_critic.parameters(), self.global_critic.parameters()
+        ):
+            global_param._grad = local_param.grad
+
+        gn = torch.nn.utils.clip_grad_norm_(
+            self.loss_module.parameters(), max_norm=max_grad_norm
+        )
+
+        self.optimizer.step()
```
Comment (theap06) [inline sota-implementations/a3c/a3c_atari.py:136]: After this line, we should sync the weights from global.

Comment (theap06) [reply sota-implementations/a3c/a3c_atari.py:136]: from_module(self.global_actor).data.to_module(self.local_actor)
from_module(self.global_critic).data.to_module(self.local_critic)

File: sota-implementations/a3c/a3c_atari.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+            self.local_actor.parameters(), self.global_actor.parameters()
+        ):
+            global_param._grad = local_param.grad
+
+        for local_param, global_param in zip(
+            self.local_critic.parameters(), self.global_critic.parameters()
+        ):
+            global_param._grad = local_param.grad
+
+        gn = torch.nn.utils.clip_grad_norm_(
```
Comment (theap06) [inline sota-implementations/a3c/a3c_atari.py:132]: I think to be inline with the correct implementation the grad clipping should occur before lines 113-118

**Review by theap06 — APPROVED** [review]

**Comment by theap06** [comment]
@simeetnayan81 lmk if you need any help with this pr

**Comment by gtnv** [comment]
hey @simeetnayan81, is this still on your list? if not, can we close it for now?
